test/functional/sessions_controller_test.rb

```diff
@@ -284,6 +367,93 @@ class SessionsControllerTest < ActionController::TestCase
     end
   end
 
+  context "#webauthn_create" do
+    context "when verifying the challenge" do
+      setup do
+        @user = create(:user)
+        @webauthn_credential = create(:webauthn_credential, user: @user)
+        post(
+          :create,
+          params: { session: { who: @user.handle, password: @user.password } }
+        )
+        @challenge = session[:webauthn_authentication]["challenge"]
+        @origin = "http://localhost:3000"
+        @rp_id = URI.parse(@origin).host
+        @client = WebAuthn::FakeClient.new(@origin, encoding: false)
+        WebauthnHelpers.create(
+          webauthn_credential: @webauthn_credential,
+          client: @client
+        )
+        post(
+          :webauthn_create,
+          params: {
+            credentials:
+              WebauthnHelpers.get_result(
+                client: @client,
+                challenge: @challenge
+              )
+          },
+          format: :json
+        )
+      end
+
+      should redirect_to :dashboard
+
+      should "log in the user" do
+        assert_predicate @controller.request.env[:clearance], :signed_in?
+      end
+    end
+
+    context "when not providing credentials" do
+      setup do
+        @user = create(:user)
+        @webauthn_credential = create(:webauthn_credential, user: @user)
+        post(
+          :create,
+          params: { session: { who: @user.handle, password: @user.password } }
+        )
+        post(
+          :webauthn_create,
+          format: :json
+        )
+      end
+
+      should respond_with :unauthorized
+    end
+
+    context "when providing wrong credentials" do
```

Comment:
I'm having trouble understanding this test -- I don't see clearly where the wrong credential is introduced.

Reply:
Ah good call out. This took me a minute to parse as well. The challenge here is a random hex, instead of being the users webauthn credential challenge, like it is above on line [380](https://github.com/rubygems/rubygems.org/pull/2865/files#diff-3330307a836699182b69bc210bd2b62dbd7ef41c26978ce7ff8e7311bd4629dcR380)

I'm renaming the variable below from `@challenge` to `@wrong_challenge` to make it clearer

Reply:
Maybe if we change the context to be something like `when credential verification fails` or `when challenge verification fails` can help with the readability here too.